pennylane/capture/autograph/transformer.py

```diff
@@ -0,0 +1,227 @@
+package (originally from TensorFlow).
+Here, we integrate AutoGraph into PennyLane to improve the UX and allow programmers to use built-in
+Python control flow and other imperative expressions rather than the functional equivalents provided
+by PennyLane.
+"""
+import copy
+import inspect
+
+from malt.core import converter
+from malt.impl.api import PyToPy
+
+import pennylane as qml
+from pennylane.capture.autograph.ag_primitives import AutoGraphError
+
+from . import ag_primitives
+
+
+class PennyLaneTransformer(PyToPy):
+    """A source-to-source transformer to convert imperative style control flow into a function style
+    suitable for tracing."""
+
+    def __init__(self):
+        super().__init__()
+
+        self._extra_locals = None
+
+    def transform(self, obj, user_context):
+        """Launch the transformation process. Typically, this only works on function objects.
+        Here we also allow QNodes to be transformed."""
+
+        # By default, AutoGraph will only convert function or method objects, not arbitrary classes
+        # such as QNode objects. Here we handle them explicitly, but we might need a more general
+        # way to handle these in the future.
+        # We may also need to check how this interacts with other common function decorators.
+        fn = obj
+        if isinstance(obj, qml.QNode):
+            fn = obj.func
+        elif inspect.isfunction(fn) or inspect.ismethod(fn):
+            pass
+        elif callable(obj):
+            # pylint: disable=unnecessary-lambda,unnecessary-lambda-assignment
+            fn = lambda *args, **kwargs: obj(*args, **kwargs)
+        else:
+            raise AutoGraphError(f"Unsupported object for transformation: {type(fn)}")
+
+        new_fn, module, source_map = self.transform_function(fn, user_context)
+        new_obj = new_fn
+
+        if isinstance(obj, qml.QNode):
+            new_obj = copy.copy(obj)
+            new_obj.func = new_fn
+
+        return new_obj, module, source_map
+
+    def get_extra_locals(self):
+        """Here we can provide any extra names that the converted function should have access to.
+        At a minimum we need to provide the module with definitions for AutoGraph primitives."""
+
+        if self._extra_locals is None:
+            extra_locals = super().get_extra_locals()
+            updates = {key: ag_primitives.__dict__[key] for key in ag_primitives.__all__}
+            extra_locals["ag__"].__dict__.update(updates)
+            self._extra_locals = extra_locals
+
+        return self._extra_locals
+
+    def has_cache(self, fn):
+        """Check for the presence of the given function in the cache. Functions to be converted are
+        cached by the function object itself as well as the conversion options."""
+
+        return (
+            self._cache.has(fn, TOPLEVEL_OPTIONS)
+            or self._cache.has(fn, NESTED_OPTIONS)
+            or self._cache.has(fn, STANDARD_OPTIONS)
```

Comment:
What are `TOP_LEVEL_OPTIONS`, etc, and where do they come from? It doesn't look like they are imported up top 🤔

Reply:
Line 210 below